tsdb/engine/tsm1/cache.go

```diff
@@ -214,28 +229,25 @@ func (c *Cache) Write(key string, values []Value) error {
 }
 
 // WriteMulti writes the map of keys and associated values to the cache. This function is goroutine-safe.
-// It returns an error if the cache has exceeded its max size.
+// It returns an error if the cache will exceeded its max size by adding the new values.
 func (c *Cache) WriteMulti(values map[string][]Value) error {
-	totalSz := 0
+	var totalSz uint64
 	for _, v := range values {
-		totalSz += Values(v).Size()
+		totalSz += uint64(Values(v).Size())
 	}
 
 	// Enough room in the cache?
-	c.mu.RLock()
-	newSize := c.size + uint64(totalSz)
-	if c.maxSize > 0 && newSize+c.snapshotSize > c.maxSize {
-		c.mu.RUnlock()
+	c.mu.Lock()
```

Comment:
Because taking the lock multiple times was causing significant slowdown due to lock contention.